### PR TITLE
Expose server's middleware for incoming clients control

### PR DESCRIPTION
**Pull request by Waldz**
Updates: https://github.com/mysteriumnetwork/node/issues/1838

- Server's callbacks to subscribe and incoming client data: Openvpn variables, states
- Expose client authorisation control

### CODE REVIEW

File: openvpn/middlewares/server/auth/middleware.go [hunk 1]
```diff
@@ -19,41 +19,89 @@ package auth
...
+// --management-client-auth directive so that it will ask the management
+// interface to approve client connections.
+type Middleware struct {
+	commandWriter management.CommandWriter
+	currentEvent  server.ClientEvent
 
-// CredentialsValidator callback checks given auth primitives (i.e. customer identity signature / node's sessionId)
-type CredentialsValidator func(clientID int, username, password string) (bool, error)
+	listenersMu sync.RWMutex
+	listeners   []ClientEventCallback
```
Comment (anjmao) [inline openvpn/middlewares/server/auth/middleware.go:42]: When are listeners removed? This could lead to memory leaks if they are never removed.

Comment (Waldz) [reply openvpn/middlewares/server/auth/middleware.go:42]: It's same for all middlewares. We simple don't reuse created instances. They are created/destroyed with server start/stop

File: openvpn/middlewares/server/credentials/middleware.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+package credentials
+
+import (
+	"github.com/mysteriumnetwork/go-openvpn/openvpn/log"
+	"github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server"
+	"github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth"
+)
+
+// Middleware is able to process authorize incoming clients by given credentials validator callback.
```
Comment (soffokl) [inline openvpn/middlewares/server/credentials/middleware.go:26]:
```suggestion
// Middleware is able to authorize incoming clients by given credentials validator callback.
```

File: openvpn/middlewares/server/credentials/middleware.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+// Validator callback checks given auth primitives.
+type Validator func(clientID int, username, password string) (bool, error)
+
+// NewMiddleware creates server user_auth challenge authentication Middleware
+func NewMiddleware(validator Validator) *Middleware {
+	m := &Middleware{
+		validator: validator,
+	}
+	m.Middleware = auth.NewMiddleware(m.handleClientEvent)
```
Comment (soffokl) [inline openvpn/middlewares/server/credentials/middleware.go:41]: I failed to parse these recursive dependencies. Can we make it more clear?
```suggestion
	m.Middleware = auth.NewMiddleware(m.handleClientEvent)
```

File: openvpn/middlewares/server/credentials/middleware.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func (m *Middleware) handleClientEvent(event server.ClientEvent) {
+	switch event.EventType {
+	case server.Connect, server.Reauth:
+		username := event.Env["username"]
+		password := event.Env["password"]
+		err := m.authenticateClient(event.ClientID, event.ClientKey, username, password)
+		if err != nil {
+			log.Error("Unable to authenticate client:", err)
```
Comment (soffokl) [inline openvpn/middlewares/server/credentials/middleware.go:52]: Are we just logging the auth problem and not returning an error?

File: openvpn/middlewares/server/credentials/middleware_test.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+type fakeValidator struct {
+	called bool
+}
+
+func (f *fakeValidator) authenticateClient(clientID int, username, password string) (bool, error) {
+	f.called = true
+	return username == "username1" && password == "12341234", nil
+}
+
+func mockAuthMiddleware(connection *management.MockConnection) *auth.Middleware {
```
Comment (soffokl) [inline openvpn/middlewares/server/credentials/middleware_test.go:38]: > func `mockAuthMiddleware` is unused (unused)

File: openvpn/middlewares/server/filter/middleware.go [hunk 6]
```diff
@@ -42,60 +42,39 @@ END
...
 var filterLAN = template.Must(template.New("filter_lan").Parse(filterLANTemplate))
 
+// Exposes API to control client's packet filtering.
+//
+// The OpenVPN server should have been started with the
+// --management-client-pf directive so that it will require that
+// VPN tunnel packets sent or received by client instances must
+// conform to that client's packet filter configuration.
 type middleware struct {
+	*auth.Middleware
```
Comment (soffokl) [inline openvpn/middlewares/server/filter/middleware.go:52]: Why `filter` depends on `auth`?

File: openvpn/middlewares/server/filter/middleware.go [hunk 7]
```diff
@@ -42,60 +42,39 @@ END
...
-		m.currentEvent.ClientID = clientID
-	case server.Env:
-		if strings.ToLower(eventData) == "end" {
-			m.handleClientEvent(m.currentEvent)
-		}
-	}
-
-	return true, nil
-}
+func (m *middleware) ConsumeLine(line string) (bool, error) { return false, nil }
```
Comment (soffokl) [inline openvpn/middlewares/server/filter/middleware.go:77]: It doesn't consume lines anymore?

File: openvpn/middlewares/server/filter/middleware_test.go [hunk 8]
```diff
@@ -21,18 +21,19 @@ import (
 	"testing"
 
 	"github.com/mysteriumnetwork/go-openvpn/openvpn/management"
+	"github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server"
 	"github.com/stretchr/testify/assert"
 )
 
 const (
-	emptyFilter = `client-pf 0
+	emptyFilter = `client-pf 3
```
Comment (soffokl) [inline openvpn/middlewares/server/filter/middleware_test.go:29]: What is the point to change it from 0 to 3?

File: openvpn/middlewares/server/credentials/middleware.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	m := &Middleware{
+		validator: validator,
+	}
+	m.Middleware = auth.NewMiddleware(m.handleClientEvent)
+	return m
+}
+
+func (m *Middleware) handleClientEvent(event server.ClientEvent) {
+	switch event.EventType {
+	case server.Connect, server.Reauth:
```
Comment (zolia) [inline openvpn/middlewares/server/credentials/middleware.go:47]: Maybe we should log other client events at least.

File: openvpn/middlewares/server/credentials/middleware.go (hunk 3 above)
Comment (Waldz) [reply openvpn/middlewares/server/credentials/middleware.go:41]: It's simple inheritance. Improved constructor

File: openvpn/middlewares/server/credentials/middleware.go (hunk 4 above)
Comment (Waldz) [reply openvpn/middlewares/server/credentials/middleware.go:52]: Did not change this logic. Probably, middleware just logs returning returned errors also.
If you mean returning error to connecting client, it's intentional. Client gets just "wrong username or password"

File: openvpn/middlewares/server/filter/middleware.go (hunk 6 above)
Comment (Waldz) [reply openvpn/middlewares/server/filter/middleware.go:52]: Simple inheritances. `auth.Middleware` general purpose - subscribes to client status events

Comment (soffokl) [reply openvpn/middlewares/server/filter/middleware.go:52]: There is no inheritance for golang, that is why it looks weird when you are trying to invent it yourself.

File: openvpn/middlewares/server/filter/middleware.go (hunk 7 above)
Comment (Waldz) [reply openvpn/middlewares/server/filter/middleware.go:77]: Yes, now subscribes for consumer client from auth.Middleware

File: openvpn/middlewares/server/filter/middleware_test.go (hunk 8 above)
Comment (Waldz) [reply openvpn/middlewares/server/filter/middleware_test.go:29]: Was triggering different events in tests. Reverted

File: openvpn/middlewares/server/credentials/middleware.go (hunk 9 above)
Comment (Waldz) [reply openvpn/middlewares/server/credentials/middleware.go:47]: This middleware does not fit in library, should be on `node` repo.
Ok, reverted this previous logging.

**Pull request merged**